### PR TITLE
LF: require type interning for 1.dev

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by hurryabit — COMMENTED** [review]
Most of the changes in `DecodeV1Spec.scala` seem unrelated. If that's the case, could we put them in a separate PR preceding this one?

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 1]
```diff
@@ -605,7 +608,21 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
...
+        uncheckedDecodeType(lfType)
+      else
+        lfType.getSumCase match {
+          case PLF.Type.SumCase.INTERNED =>
+            internedTypes.applyOrElse(
+              lfType.getInterned,
+              (index: Int) => throw ParseError(s"invalid internedTypes table index $index"),
+            )
+          case otherwise =>
+            throw ParseError(s"$otherwise is not supported outside type interning table")
```
Comment (hurryabit) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:622]: Just to confirm I understand this correctly: In DAML-LF 1.dev every type outside of the type interning table has to be a reference into the type interning table and types inside the type interning table can be arbitrarily complex and contain references into the table at any level.

**Review by hurryabit — APPROVED** [review]
This is great work. Thank you!

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:622]: Yes. 
* `decodeType` enforces (for 1.dev only) that `lfType` is a reference to the the type interning table. (should be use to decode type outside the interning table)
* `uncheckedDecodeType` allows `lfType` to be arbitrarily complex and contain references into the table at any level.  (should be use to decode type inside the interning table)

Comment (hurryabit) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:622]: Thanks for confirming.

**Comment by remyhaemmerle-da** [comment]
> Most of the changes in `DecodeV1Spec.scala` seem unrelated. If that's the case, could we put them in a separate PR preceding this one?

I created a separated PR. Could you please double check  `DecodeV1Spec.scala` that contains now only "related" changes.

**Review by hurryabit — APPROVED** [review]
Very nice. Thank you!

And thanks a lot for moving the refactoring of the decoder tests into a separate PR.

**Pull request merged**